include/lbann/callbacks/callback.hpp

```diff
@@ -89,6 +90,10 @@ class callback_base {
   /** @name Callback hooks */
   ///@{
 
+  /** @brief Called at the beginning of setup. */
+  virtual void on_setup_begin(model *m) {}
```

Comment:
Also, I think `on_setup_begin(model *m)` is not a valid hook -- the callback may not have been setup yet since callback setup is done as part of model setup... While many callbacks don't have a nonempty setup function, this is a semantic issue.

Reply:
Yea, I was thinking about whether this was a good idea or not. I can imagine use-cases (printing information about the LBANN build or just printing the fact that setup has begun), but perhaps it's just too dangerous.

Reply:
"Dangerous" might be overselling, but certainly its potential for semantic nonsense is quite high. Also, things like the build information are not characteristic of a single model but are truly global. There should be another, independent mechanism for reporting this information. The `print_model_description` callback makes semantic sense because callbacks are model-specific and it just prints the specific model's description.